samza-core/src/main/java/org/apache/samza/table/retry/AsyncRetriableTable.java

```diff
@@ -0,0 +1,158 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.table.retry;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+
+import java.util.function.Predicate;
+import net.jodah.failsafe.RetryPolicy;
+
+import org.apache.samza.context.Context;
+import org.apache.samza.storage.kv.Entry;
+import org.apache.samza.table.AsyncReadWriteTable;
+import org.apache.samza.table.remote.TableReadFunction;
+import org.apache.samza.table.remote.TableWriteFunction;
+import org.apache.samza.table.utils.TableMetricsUtil;
+
+import static org.apache.samza.table.retry.FailsafeAdapter.failsafe;
+
+import static org.apache.samza.table.BaseReadWriteTable.Func1;
+
+
+/**
+ * A composable asynchronous retriable table implementation that supports features
+ * defined in {@link TableRetryPolicy}.
+ *
+ * @param <K> the type of the key in this table
+ * @param <V> the type of the value in this table
+ */
+public class AsyncRetriableTable<K, V> implements AsyncReadWriteTable<K, V> {
+
+  private final String tableId;
+  private final AsyncReadWriteTable<K, V> table;
+  private final RetryPolicy readRetryPolicy;
+  private final RetryPolicy writeRetryPolicy;
+  private final ScheduledExecutorService retryExecutor;
+
+  @VisibleForTesting
+  RetryMetrics readRetryMetrics;
+  @VisibleForTesting
+  RetryMetrics writeRetryMetrics;
+
+  public AsyncRetriableTable(String tableId, AsyncReadWriteTable<K, V> table,
+      TableRetryPolicy readRetryPolicy, TableRetryPolicy writeRetryPolicy, ScheduledExecutorService retryExecutor,
+      TableReadFunction readFn, TableWriteFunction writeFn) {
+
+    Preconditions.checkNotNull(tableId, "null tableId");
+    Preconditions.checkNotNull(table, "null table");
+    Preconditions.checkNotNull(retryExecutor, "null retryExecutor");
+    Preconditions.checkArgument(readRetryPolicy != null || writeRetryPolicy != null,
+        "both readRetryPolicy and writeRetryPolicy are null");
+
+    this.tableId = tableId;
+    this.table = table;
+    this.retryExecutor = retryExecutor;
+
+    if (readRetryPolicy != null && readFn != null) {
+      Predicate<Throwable> readRetryPredicate = readRetryPolicy.getRetryPredicate();
+      readRetryPolicy.withRetryPredicate((ex) -> readFn.isRetriable(ex) || readRetryPredicate.test(ex));
+      this.readRetryPolicy = FailsafeAdapter.valueOf(readRetryPolicy);
+    } else {
+      this.readRetryPolicy = null;
```

Comment:
can the "else" part be removed since by default the object variable of a class is null?
Is it possible to refactor the code for readRetryPolicy and writeRetryPolicy since they are very similar?

Reply:
They are final variables, so they would need a value assignment.
Can you elaborate more re readRetryPolicy and writeRetryPolicy, the intent is to provide different retry behavior for read and write.